### PR TITLE
Ipf adjustment fixes

### PR DESCRIPTION
**Pull request by isaac-tfn**
Describe Changes
---

Task Checklist
----
- [ ] Have unittests been added (testing individual functions and their edge cases)
- [ ] Have integration tests been added (if applicable, to test modules of functionality)
- [ ] Updated RELEASE.md to reflect changes in PR
- [ ] Have new dependencies been added?
  - [ ] Have they been added to either `requirements.txt` or `requirements_dev.txt`.
  - [ ] Have they been added to `setup.cfg`

### CODE REVIEW

File: src/caf/base/data_structures.py [hunk 1]
```diff
@@ -1536,18 +1572,43 @@ def validate_ipf_targets(
...
-                target_sum = target.data.sum()
+            subsets = target.data.segmentation.input.subsets
+            if len(subsets) > 0:
+                for comp_target in targets:
+                    if target == comp_target:
+                        continue
+                    # Check that subsets exist in comp_target
+                    if set(subsets.keys()).intersection(
+                        comp_target.data.segmentation.names
+                    ) != set(subsets.keys()):
```
Comment (MattBuckley-TfN) [inline src/caf/base/data_structures.py:1583]: I think this could be rewritten as a set comparison which checks if every element in set on the left is in set on the right
```suggestion
                    if set(subsets.keys()) <= set(
                        comp_target.data.segmentation.names
                    ):
```

File: src/caf/base/data_structures.py [hunk 2]
```diff
@@ -1536,18 +1572,43 @@ def validate_ipf_targets(
...
-                        "Input target DVectors do not have consistent "
-                        f"sums, so ipf will fail target at position {position} doesn't match "
-                        "the first target. It is possible later targets also don't match."
-                    )
+                # Check targets sum to the same, or they can't converge. Potentially could allow
+                # IPF for non-agreeing targets to get as close as possible.
+                if target_sum == 0:
+                    target_sum = target.data.sum()
+                else:
+                    # TODO don't hard code this
```
Comment (MattBuckley-TfN) [inline src/caf/base/data_structures.py:1605]: This comment has been fixed
```suggestion
```

File: src/caf/base/data_structures.py [hunk 3]
```diff
@@ -2201,7 +2262,107 @@ class IpfTarget:
     )
 
     @staticmethod
-    def check_compatibility(targets: Collection[IpfTarget], adjust: bool = False):
+    def _check_loop(target_1, target_2, adjust, targ_dict, ind, rmses, trans_cache=None):
```
Comment (MattBuckley-TfN) [inline src/caf/base/data_structures.py:2265]: Type annotations and a brief docstring would help understand the method it would be useful to know the justification for skipping checks in some cases e.g. when target_2 segmentation has subsets.

File: src/caf/base/data_structures.py [hunk 4]
```diff
@@ -2201,7 +2262,107 @@ class IpfTarget:
...
+                adj = adj.replace(to_replace={np.inf: 0})
+                target_1.data = target_1.data.mul(adj, axis=1)
+            else:
+                if not isinstance(adj, Number):
+                    adj.fill(np.inf, 0)
+                    target_1 *= adj
+            targ_dict[ind] = target_1
+        return targ_dict, rmses
+
+    @staticmethod
```
Comment (MattBuckley-TfN) [inline src/caf/base/data_structures.py:2358]: This is a classmethod not a staticmethod because it uses other methods (`_check_loop`) from the IpfTarget class

File: src/caf/base/data_structures.py [hunk 5]
```diff
@@ -2218,27 +2379,37 @@ def check_compatibility(targets: Collection[IpfTarget], adjust: bool = False):
         adjust: bool = False
             Whether to change the targets or just report on their compatibility.
         """
+        targets = list(targets)
         targ_dict = {i: j for i, j in enumerate(targets)}
```
Comment (MattBuckley-TfN) [inline src/caf/base/data_structures.py:2383]: Doesn't need changing but I don't understand the point of changing targets into a dictionary and just using the index as keys, you could just leave it as a list a use indices in the same way.

File: src/caf/base/data_structures.py [hunk 6]
```diff
@@ -2218,27 +2379,37 @@ def check_compatibility(targets: Collection[IpfTarget], adjust: bool = False):
         adjust: bool = False
```
Comment (MattBuckley-TfN) [inline src/caf/base/data_structures.py:2379]: Missing explainations for reference, chain_adjust and trans_cache parameters

File: src/caf/base/data_structures.py [hunk 7]
```diff
@@ -2201,7 +2262,107 @@ class IpfTarget:
...
+                        agg_2.zoning_system, cache_path=trans_cache
+                    )
+                nested_1 = (
+                    trans[agg_1.zoning_system.translation_column_name(agg_2.zoning_system)]
+                    == 1
+                ).all()
+                nested_2 = (
+                    trans[agg_2.zoning_system.translation_column_name(agg_1.zoning_system)]
+                    == 1
+                ).all()
```
Comment (MattBuckley-TfN) [inline src/caf/base/data_structures.py:2300]: Having these across multiple lines is quite difficult to read might be easier if `agg_1.zoning_system...` was set to a variable so the the nested line could be simpler e.g. `nested_1 = (trans[agg_variable] == 1).all()`

**Review by MattBuckley-TfN — CHANGES_REQUESTED** [review]
Logic all looks good, just some comments on docstrings and type annotations and a few clarification questions.

File: src/caf/base/data_structures.py [hunk 8]
```diff
@@ -345,8 +346,8 @@ def data(self, value: pd.Series | pd.DataFrame):
             raise TypeError(
                 "data must be a pandas DataFrame or Series. Input " f"value is {value.type}."
             )
-        if isinstance(value, pd.Series):
-            value = value.to_frame()
+        # if isinstance(value, pd.Series):
+        #     value = value.to_frame()
```
[the remark this answers is not in the document]
Comment (isaac-tfn) [reply src/caf/base/data_structures.py:350]: Yes the type annotation already did allow series so this should have already been removed

File: src/caf/base/data_structures.py (hunk 5 above)
Comment (isaac-tfn) [reply src/caf/base/data_structures.py:2382]: Good point

File: src/caf/base/segments/economic_status.yml [hunk 9]
```diff
@@ -20,4 +20,21 @@ exclusions:
       - 4
       - 7
       - 10
+lookups:
```
[the remark this answers is not in the document]
Comment (isaac-tfn) [reply src/caf/base/segments/economic_status.yml:23]: Yep will message you

File: src/caf/base/zoning.py [hunk 10]
```diff
@@ -555,6 +555,37 @@ def _replace_id(
 
         return translation
 
+    def _check_all_columns(self, input_columns):
```
[the remark this answers is not in the document]
Comment (isaac-tfn) [reply src/caf/base/zoning.py:558]: I'll make it public and add a docstring, but I'm not going to change to return None, as I don't want to track where it may return that all over the place.

**Review by isaac-tfn — COMMENTED** [review]
Made changes suggested, once I push up I expect to need a couple more tweaks for linting to pass

**Review by MattBuckley-TfN — APPROVED** [review]
